artifacts/writer.py

```diff
@@ -0,0 +1,93 @@
+# -*- coding: utf-8 -*-
+"""The artifact writer objects."""
+
+import abc
+import json
+import yaml
+import sys
+from artifacts.artifact import ArtifactDefinition
+
+
+class BaseArtifactsWriter(object):
+  """Class that implements the artifacts reader interface."""
+
+  @abc.abstractmethod
+  def WriteArtifactsFile(self, artifacts, filename):
+    """Writes artifact definitions to a file.
+
+    Args:
+      artifacts: list of ArtifactDefinition objects to be written
+
+      filename: filename to write artifacts to
+    """
+
+  @abc.abstractmethod
+  def FormatArtifacts(self, artifacts):
+    """Formats artifacts to desired output format
+
+    Args:
+      artifacts: an instance or list of ArtifactDefinition
+
+    Returns:
+      formatted string of artifact_definition
+    """
+
+
+class ArtifactWriter(BaseArtifactsWriter):
+
+  def WriteArtifactsFile(self, artifacts, filename):
+    """Writes artifact definitions to a file.
+
+    Args:
+      artifacts: list of ArtifactDefinition objects to be written
+
+      filename: filename to write artifacts to
+    """
+    with open(filename, 'wb') as file_object:
+      file_object.write(self.FormatArtifacts(artifacts))
+
+
+class JsonArtifactsWriter(ArtifactWriter):
+
+  def FormatArtifacts(self, artifacts):
+    """Converts artifacts to dict from ArtifactDefinition objects
+
+    Args:
+      artifacts: an instance or list of ArtifactDefinition
+
+    Returns:
+      formatted string of artifact_definition
+    """
+    if isinstance(artifacts, ArtifactDefinition):
```

Comment:
That is what I mean by an instance or list of ArtifactDefinition.  If you pass an instance it turns it into a list similar to how GRR handles similar cases.  I can change the docstring to be more clear or just force it to only accept lists.